File: libraries/java/src/main/java/com/qmplus/v3/api/services/LDAPService.java
package com.qmplus.v3.api.services;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.qmplus.v3.api.models.request.*;
import com.qmplus.v3.api.models.response.*;
import com.qmplus.v3.api.models.vo.VoMessage;

import java.io.IOException;
import java.util.List;
import java.util.Properties;

public class LDAPService extends BaseService {

  public LDAPService() {
    super();
  }

  public LDAPService(String endPoint) {
    super(endPoint);
  }

  /**
   * Retrieve LDAP status for the current tenant
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPStatusResponse status(
    String authTokenKey, String tenant, String clientInfo
  ) throws IOException {
    LDAPStatusRequest request = new LDAPStatusRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");

    // Execute the operation
    ResponseWrapper<LDAPStatusResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPStatusResponse>>() {},"ldap/status", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP attributes for a user
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @param username The username we are trying to look up
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPUserResponse ldapUser(
    String authTokenKey, String tenant, String clientInfo, String username
  ) throws IOException {
    LDAPUserRequest request = new LDAPUserRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setUsername(username);

    // Execute the operation
    ResponseWrapper<LDAPUserResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPUserResponse>>() {},"ldap/user/ldap", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP information for a given department by department id.
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @param departmentId The department id.
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPDepartmentInfoResponse departmentInfoById(
    String authTokenKey, String tenant, String clientInfo, Integer departmentId
  ) throws IOException {
    LDAPDepartmentInfoRequest request = new LDAPDepartmentInfoRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setId(departmentId);

    // Execute the operation
    ResponseWrapper<LDAPDepartmentInfoResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPDepartmentInfoResponse>>() {},"ldap/department/info", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP information for a given department by department name.
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @param departmentName The department name.
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPDepartmentInfoResponse departmentInfoByName(
    String authTokenKey, String tenant, String clientInfo, String departmentName
  ) throws IOException {
    LDAPDepartmentInfoRequest request = new LDAPDepartmentInfoRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setName(departmentName);

    // Execute the operation
    ResponseWrapper<LDAPDepartmentInfoResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPDepartmentInfoResponse>>() {},"ldap/department/info", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP results from the departments LDAP filters
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @param departmentId The department id.
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPDepartmentQueryResponse departmentQueryById(
    String authTokenKey, String tenant, String clientInfo, Integer departmentId
  ) throws IOException {
    LDAPDepartmentQueryRequest request = new LDAPDepartmentQueryRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setId(departmentId);

    // Execute the operation
    ResponseWrapper<LDAPDepartmentQueryResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPDepartmentQueryResponse>>() {},"ldap/department/query", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP results from the departments LDAP filters
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @param departmentName The department name.
   * @return Returns a LDAPSettingsResponse object for a successful status query.
   * @throws IOException Throws an IOException.
   */
  public LDAPDepartmentQueryResponse departmentQueryByName(
    String authTokenKey, String tenant, String clientInfo, String departmentName
  ) throws IOException {
    LDAPDepartmentQueryRequest request = new LDAPDepartmentQueryRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setName(departmentName);

    // Execute the operation
    ResponseWrapper<LDAPDepartmentQueryResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPDepartmentQueryResponse>>() {},"ldap/department/query", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  /**
   * Retrieve LDAP settings for the current tenant
   * Logged in user required LDAP system roles
   *
   * @param authTokenKey The auth token of the user we are going to log out.
   * @param tenant The name of the tenant.
   * @param clientInfo A text string representing information about the calling client.
   * @return Returns a LDAPSettingsResponse object for a successful setting retrieval.
   * @throws IOException Throws an IOException.
   */
  public LDAPSettingsResponse settings(
    String authTokenKey, String tenant, String clientInfo
  ) throws IOException {
    LDAPSettingsRequest request = new LDAPSettingsRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");

    // Execute the operation
    ResponseWrapper<LDAPSettingsResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPSettingsResponse>>() {},"ldap/settings", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }

  public LDAPQueryResponse query(
    String authTokenKey, String tenant, String clientInfo,
    String searchBase, List<String> filters
  ) throws IOException {
    LDAPQueryRequest request = new LDAPQueryRequest();

    // Trigger the default random key
    request.setAuthTokenKey(authTokenKey);
    request.setTenant(tenant);
    request.setClientInfo(clientInfo != null ? clientInfo : "");
    request.setSearchBase(searchBase);
    request.setFilters(filters);

    // Execute the operation
    ResponseWrapper<LDAPQueryResponse> response
      = executeOperation(new TypeReference<ResponseWrapper<LDAPQueryResponse>>() {},"ldap/query", toJson(request));

    // Return the content object if everything went well
    return response.getContent();
  }
}

File: utilities/ldap-cli/lib/ldap_service.js
const BaseService = require('./service');

class LDAPService extends BaseService {
  constructor(endpoint) {
    super(endpoint);
  }

  /**
   * Perform an ad-hoc query to the configured LDAP servers allowing you to test queries.
   * @param authTokenKey The user's authTokenKey
   * @param tenant The user's tenant (organization name)
   * @param searchbase The LDAP query searchbase.
   * @param filters A list of filters to apply to the query.
   * @returns Status
   */
  async query(authTokenKey, tenant, searchbase, filters) {
    let result = await this.executeOperation("/ldap/query", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      searchBase: searchbase,
      filters: filters
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to execute the ldap query for tenant ${tenant}`);
    }    
  }

  /**
   * Retrieve the ldap status for a tenant
   * @param authTokenKey The user's authTokenKey
   * @param tenant The user's tenant (organization name)
   * @returns Status
   */
  async status(authTokenKey, tenant) {
    let result = await this.executeOperation("/ldap/status", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile"
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant}`);
    }    
  }

  /**
   * Retrieve the ldap settings for a tenant
   * @param authTokenKey The user's authTokenKey
   * @param tenant The user's tenant (organization name)
   * @returns Status
   */
  async settings(authTokenKey, tenant) {
    let result = await this.executeOperation("/ldap/settings", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile"
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap settings for tenant ${tenant}`);
    }    
  }

  /**
   * Retrieve LDAP attributes for a user
   * @param authTokenKey The user's authTokenKey
   * @param tenant The tenant (organization name)
   * @param username The username we want to check
   * @returns UserStatus
   */
  async ldapUser(authTokenKey, tenant, username) {
    let result = await this.executeOperation("/ldap/user/ldap", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      username: username
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant} and user ${username}`);
    }    
  }

  /**
   * Retrieve LDAP information for a given department by department id.
   * @param authTokenKey The user's authTokenKey
   * @param tenant The tenant (organization name)
   * @param departmentId The departmentId we want to check
   * @returns UserStatus
   */
  async departmentInfoById(authTokenKey, tenant, departmentId) {
    let result = await this.executeOperation("/ldap/department/info", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      id: departmentId
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant} and department ${departmentId}`);
    }    
  }

  /**
   * Retrieve LDAP information for a given department by department name.
   * @param authTokenKey The user's authTokenKey
   * @param tenant The tenant (organization name)
   * @param name The name we want to check
   * @returns UserStatus
   */
  async departmentInfoByName(authTokenKey, tenant, name) {
    let result = await this.executeOperation("/ldap/department/info", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      name: name
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant} and department ${name}`);
    }    
  }

  /**
   * Retrieve LDAP results from the departments LDAP filters by department id.
   * @param authTokenKey The user's authTokenKey
   * @param tenant The tenant (organization name)
   * @param departmentId The departmentId we want to check
   * @returns UserStatus
   */
  async departmentQueryById(authTokenKey, tenant, departmentId) {
    let result = await this.executeOperation("/ldap/department/query", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      id: departmentId
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant} and department ${departmentId}`);
    }    
  }

  /**
   * Retrieve LDAP results from the departments LDAP filters by department name.
   * @param authTokenKey The user's authTokenKey
   * @param tenant The tenant (organization name)
   * @param name The departmentId we want to check
   * @returns UserStatus
   */
  async departmentQueryByName(authTokenKey, tenant, name) {
    let result = await this.executeOperation("/ldap/department/query", {
      authTokenKey: authTokenKey,
      tenant: tenant,
      clientInfo: "mobile",
      name: name
    });

    if (result) {
      return result.content;
    } else {
      throw Error(`failed to retrieve ldap status for tenant ${tenant} and department ${name}`);
    }    
  }
}

module.exports = LDAPService;